flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaSink.java

```diff
@@ -86,45 +85,48 @@
         return new KafkaSinkBuilder<>();
     }
 
+    @Internal
```

Comment:
The problem here is that we do not differentiate between implementation internal and usage internal. The method `createWriter` etc. are usage internal but **not** implementation internal. So developers need to implement them but none of the users should use them because they are called by the Flink runtime.

From now on every connector developer needs to annotate the method that users should not use directly.

Only introducing a second `@Internal` annotation would solve the problem completely IMO but I do not want to block this PR with that discussion.